### PR TITLE
[CPU] PagedAttention supports dynamic-split fuse

### PR DESCRIPTION
**Pull request by luo-cheng2021**
### Details:
 - *Merge first token and second token inference into one parallel loop*
 - *~~Additional optimization: pre-transpose k-cache, pre-pack v-cache if needed~~*
 - *Additional optimization for first token: save q * k' upper triangle matrix computation and (q * k') * v lower triangle matrix computation* 
 - *C++ pipeline can enable it: https://github.com/ilya-lavrenov/openvino.genai/pull/9*
 - *TODO(in another PR):*
   - alibi support
   - performance tuning
   - testcase

### Tickets:
 - *[138673](https://jira.devtools.intel.com/browse/CVS-138673)*

### CODE REVIEW

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/transpose_kernel.hpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    _mm512_storeu_si512(dst + 9 * dst_stride, r9);
+    _mm512_storeu_si512(dst + 10 * dst_stride, ra);
+    _mm512_storeu_si512(dst + 11 * dst_stride, rb);
+    _mm512_storeu_si512(dst + 12 * dst_stride, rc);
+    _mm512_storeu_si512(dst + 13 * dst_stride, rd);
+    _mm512_storeu_si512(dst + 14 * dst_stride, re);
+    _mm512_storeu_si512(dst + 15 * dst_stride, rf);
+}
+
+inline void transpose_16x16_kernel(uint32_t* dst, uint32_t* src, size_t dst_stride, size_t src_stride) {
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/transpose_kernel.hpp:136]: What will happen if input data is not aligned by 16 * 16 ? Maybe a comment should be added if this function requires memory aligned

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp [hunk 2]
```diff
@@ -134,9 +140,14 @@ void BrgemmKernel::init_brgemm(brgemmCtx& ctx,
...
-    auto isa = use_amx                                     ? isa_undef
-               : ctx.dt_in0 == dnnl_data_type_t::dnnl_bf16 ? avx512_core_bf16
-                                                           : (is_int8 ? avx512_core_vnni : avx512_core);
+    cpu_isa_t isa;
+    if (mayiuse(avx512_core)) {
+        isa = use_amx ? isa_undef
+                : ctx.dt_in0 == dnnl_data_type_t::dnnl_bf16 ? avx512_core_bf16
+                                                            : (is_int8 ? avx512_core_vnni : avx512_core);
+    } else {
+        isa = cpu_isa_t::avx2;
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp:149]: Is AVX2 a must need for PA ? When will AVX2 be selected ?

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp [hunk 3]
```diff
@@ -158,6 +169,19 @@ void BrgemmKernel::init_brgemm(brgemmCtx& ctx,
         THROW_ERROR("cannot be executed due to invalid brgconv params");
     }
 
+    if (use_amx && b_accumulate) {
+        brgemm_attr_t brgattr;
+        brgattr.max_bs = 1;
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp:174]: Could you add comments why we need this configuration ?

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 4]
```diff
@@ -0,0 +1,1595 @@
...
+                _helper._S, _helper._block_size, _helper._S);
+            if (q_is_bf16) {
+                pack_32Nx32K(_helper._wv_scratch_b.template ptr<DATA_TYPE>(b, kv_block, hk),
+                    v_ptr,
+                    _helper._output.template ptr<DATA_TYPE>(ithr),
+                    _helper._block_size,
+                    _helper._S,
+                    _helper._S);
+            } else {
+                if (!q_cache_is_same) {
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1163]: Maybe you should clarify the `else` condition. The `else` branch means query is not bf16 and the query's precision is not same with KV_cache. When will this happen, when the query is `fp32` while KV cache is stored in `U8` ?

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/attn_quant_kernel.hpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+template<typename TDST>
+void attn_dequant_u8_kernel(const uint8_t* src, TDST* dst, size_t n, float scale, float zp) {
+    size_t i = 0;
+    // loadu_si128/epi64 does not support const qualifier
+    uint8_t* src_nc = const_cast<uint8_t*>(src);
+#if defined(HAVE_AVX512F)
+    auto v_zp = _mm512_set1_ps(zp);
+    auto v_scale = _mm512_set1_ps(scale);
+    for (; i + vec_len_f32_avx512 <= n; i += vec_len_f32_avx512) {
```
[the remark this answers is not in the document]
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/attn_quant_kernel.hpp:26]: Will try in later performance tuning, thanks.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/transpose_kernel.hpp (hunk 1 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/transpose_kernel.hpp:136]: We checked in  block_size must be multiple of 32 and head_size must be multiple of 16 in [AttentionExecutor::execute ](https://github.com/openvinotoolkit/openvino/blob/c23e096b4289bfadb04435d19eca94894135a86a/src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp#L1508-L1509) which should satisfy the [llm model needs](https://github.com/ilya-lavrenov/vllm/blob/2e5648ab25a0e889f4212ef74530f01ebf638dcd/vllm/model_executor/layers/attention/ops/paged_attn.py#L15), so we can safely use 16x16 kernels.

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp (hunk 2 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp:149]: AVX2 may use for testing. Not sure real case needs it.

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp (hunk 3 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp:174]: Comments added.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 4 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1187]: Yes, comments added.

**Comment by ilya-lavrenov** [comment]
@zhangYiIntel @dmitry-gorokhov 
Could you please review once again and merge PR if no comments? 
We want to merge this PR rather sooner than later to migrate to new PA interface and unlock other work.

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp (hunk 3 above)
Comment (zhangYiIntel) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp:174]: Does this has side effect when the machine has no AMX support ?

Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp:174]: No, non-amx will use `vaddps` to complete accumulation.

**Review by zhangYiIntel — APPROVED** [review]
LGTM, thanks!

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**